src/CircularArrays.jl

```diff
@@ -26,19 +26,35 @@ Alias for [`CircularArray{T,1}`](@ref).
 """
 const CircularVector{T} = CircularArray{T, 1}
 
-@inline clamp_bounds(arr::CircularArray, I::Tuple{Vararg{Int}})::AbstractArray{Int, 1} = map(dim -> mod1(I[dim], size(arr.data, dim)), eachindex(I))
+# Copied from a method of Base.mod, for compatibility with Julia version < 1.3,
+# where this method is not defined
+_mod(i::Integer, r::AbstractUnitRange{<:Integer}) = mod(i-first(r), length(r)) + first(r)
+@inline clamp_bounds(arr::CircularArray, I::Tuple{Vararg{Int}})::AbstractArray{Int, 1} = map(Base.splat(_mod), zip(I, axes(arr.data)))
 
 CircularArray(def::T, size) where T = CircularArray(fill(def, size))
 
-@inline Base.getindex(arr::CircularArray, i::Int) = @inbounds getindex(arr.data, mod1(i, size(arr.data, 1)))
-@inline Base.setindex!(arr::CircularArray, v, i::Int) = @inbounds setindex!(arr.data, v, mod1(i, size(arr.data, 1)))
+@inline Base.getindex(arr::CircularArray, i::Int) = @inbounds getindex(arr.data, mod(i, Base.axes1(arr.data)))
+@inline Base.setindex!(arr::CircularArray, v, i::Int) = @inbounds setindex!(arr.data, v, mod(i, Base.axes1(arr.data)))
 @inline Base.getindex(arr::CircularArray, I::Vararg{Int}) = @inbounds getindex(arr.data, clamp_bounds(arr, I)...)
 @inline Base.setindex!(arr::CircularArray, v, I::Vararg{Int}) = @inbounds setindex!(arr.data, v, clamp_bounds(arr, I)...)
 @inline Base.size(arr::CircularArray) = size(arr.data)
+@inline Base.axes(arr::CircularArray) = axes(arr.data)
+
+@inline Base.checkbounds(::CircularArray, _...) = nothing
+
+@inline _similar(arr::CircularArray, ::Type{T}, dims) where T = CircularArray(similar(arr.data,T,dims))
+@inline Base.similar(arr::CircularArray, ::Type{T}, dims::Tuple{Base.DimOrInd, Vararg{Base.DimOrInd}}) where T = _similar(arr,T,dims)
```

Comment:
After some more testing, I noticed that this line does not seem to be covered by the tests. Do you happen to have an example for where it might be used? Or did the addition of the ambiguity resolution for OffsetArrays cover this? Technically, this method is more broad than the one below since DimOrInd is more generic, but I cannot think of anything that would ever use this, short of implementing your axis type, which would create the same ambiguity as below anyway.

Reply:
Yes, I added this before the ambiguity resolution.
I think you're right, and we will face an ambiguity error on `similar` when indexing with arrays defining a custom axis type in the "standard" way defined in the Julia docs.  I tested this with `CatIndices.BidirectionalVector`, and indeed indexing `c[bidi_vec]` where `c` is a `CircularArray` triggers the ambiguity. If this more general `similar` method is removed the ambiguity disappears, but the result of indexing is no longer circular, which is inconsistent with `c[a::Array]`.
I think maintaining the circularity in, e.g. `c[1:5]` is really useful, and it's unfortunate if we can't have it for `c[i]` where `i` has custom axes, but I'm not sure how it would be possible with how array indexing is designed in Base.

Reply:
I guess I will keep this as it is for now.